fxa-oauth-server/lib/routes/token.js

```diff
@@ -139,221 +146,199 @@ module.exports = {
+    if (! pkceHashValue) {
+      throw AppError.missingPkceParameters();
     }
-    const pkceHashValue = pkceHash(pkceVerifier);
-    if (codeObj.codeChallenge &&
-        codeObj.codeChallengeMethod === 'S256' &&
-        pkceHashValue === codeObj.codeChallenge) {
-      return db.removeCode(buf(code)).then(function() {
-        return codeObj;
-      });
-    } else {
+    // Did they provide the *correct* challenge verifier?
+    if (! crypto.timingSafeEqual(Buffer.from(codeObj.codeChallenge), Buffer.from(pkceHashValue))) {
       throw AppError.mismatchCodeChallenge(pkceHashValue);
     }
-  });
+  }
+  // Is a very confused client is attempting to use PCKE to claim a code
+  // that wasn't actually created using PKCE?
+  if (params.code_verifier && ! codeObj.codeChallenge) {
+    throw AppError.mismatchCodeChallenge(pkceHashValue);
+  }
+  // Looks legit! Codes are one-time-use, so remove it from the db.
+  await db.removeCode(buf(code));
+  return codeObj;
 }
 
-function confirmClientSecret(client, secret) {
-  return P.resolve().then(function() {
-    var id = client.id;
-    var submitted = hex(encrypt.hash(buf(secret)));
-    var stored = hex(client.hashedSecret);
-
-    if (submitted !== stored) {
-      var storedPrevious;
-      if (client.hashedSecretPrevious) {
-        // Check if secret used is the current previous secret
-        storedPrevious = hex(client.hashedSecretPrevious);
-        if (submitted === storedPrevious) {
-          logger.info('client.matchSecretPrevious', { client: id });
-          return client;
-        }
-      }
-
-      logger.info('client.mismatchSecret', { client: id });
-      logger.verbose('client.mismatchSecret.details', {
-        submitted: submitted,
-        db: stored,
-        dbPrevious: storedPrevious
-      });
-      throw AppError.incorrectSecret(id);
-    }
-
-    return client;
-  });
-}
 
-function confirmCode(id, code) {
-  return db.getCode(buf(code)).then(function(codeObj) {
-    if (! codeObj) {
-      logger.debug('code.notFound', { code: code });
-      throw AppError.unknownCode(code);
-    } else if (hex(codeObj.clientId) !== hex(id)) {
-      logger.debug('code.mismatch', {
-        client: hex(id),
-        code: hex(codeObj.clientId)
-      });
-      throw AppError.mismatchCode(code, id);
-    } else {
-      // + because loldatemath. without it, it does string concat
-      var expiresAt = +codeObj.createdAt + config.get('expiration.code');
-      if (Date.now() > expiresAt) {
-        logger.debug('code.expired', { code: code });
-        throw AppError.expiredCode(code, expiresAt);
-      }
-    }
-    return db.removeCode(buf(code)).then(function() {
-      return codeObj;
+async function validateRefreshTokenGrant(client, params) {
+  // Does the refresh token actually exist?
+  const tokObj = await db.getRefreshToken(encrypt.hash(params.refresh_token));
+  if (! tokObj) {
+    logger.debug('refresh_token.notFound', params.refresh_token);
```

Comment:
For Stackdriver to injest as expected, does the 2nd parameter to all logger methods need to be an object? cc @jbuck

Reply:
I'll fix them for consistency, even if they might haved worked as-is.

Reply:
How does mozlog transform it? Without knowing that, I'd assume it's safer to convert it into an object yourself